src/dodal/util.py

```diff
@@ -0,0 +1,16 @@
+from importlib import import_module
+from types import ModuleType
+from typing import Any, Callable, Dict, Iterable, Union
+
+
+def make_all_devices(module: Union[str, ModuleType, None] = None) -> Dict[str, Any]:
```

Comment:
This is just an idea but I think it would be useful to have _some_ way of creating all the devices in a beamline module. The approach here is just to take all functions in a module whose names begin with `make_` (inspired by various test frameworks). An alternative is some sort of `@device` decorator to mark functions to be called. Either way would also have to decide whether to allow functions that take non-default arguments and what to do with them if so.

Reply:
I agree that something like that is nice, makes https://github.com/DiamondLightSource/dodal/issues/3 easier too. As above, I'm not sure if the `make_` convention is 100% correct anyway though. Could we inspect if the return type is a device and use that?

Reply:
Ooh, that's an idea! It does require the user to remember to include a return type in the function signature. Do we prefer that to the generator? I guess it's not _added_ boilerplate per-say as we generally expect annotated code in these repos. Some boilerplate is required for all 3 options so relative pros are:

* `make_`: similar to pytest et al. so may be familiar to the user
* Inspect return type: the user was probably going to include that anyway
* Decorator: explicit, makes it very obvious to the user which functions will get picked up

I have no strong attachment to any of these, I'm all ears for opinions.

Reply:
For v2, we could put all this in the `__init__` function of the compound devices

```python
class Choppers(StandardReadable):
    """
    Grouping for the P45 chopper motors
    """
    def __init__(self, name="choppers", prefix=f"{_settings.pv_prefix}-MO-CHOP-01:"):
        self.x = EpicsMotor("ENDAT")
        self.y = EpicsMotor("BISS")
        super().__init__(prefix, name)
```

Although I'm not convinced about the `make_` functions in general, in my experiences beamlines all have subtle differences, and I prefer the readability of instantiating all the devices once per beamline to the DRY of some `make_` functions:

```python
with DeviceCollector(set_name=True):
    choppers = p45.Choppers()
    det = p45.AdAravisDetector()
```

Reply:
Ultimately, this approach comes from offline discussions in which we agreed we were against imports with side effects and nothing should be created just by importing the module. Is it possible to make an approach like this work with that constraint?

Reply:
(wrote a version that inspects the return type, just to see how it looks)

Reply:
I'm not suggesting making a module with side effects, I'm suggesting putting all the device creation in a beamline startup script, and if there is duplication between beamlines then so be it. However I am aware this comes from a non-MX experience where every beamline is a bit different, so maybe the MX use-case is strong enough for `make_` functions. In this case though, could you just have an MX specific `make_` function rather than requiring it as an approach for everything?

Reply:
I don't think MX are completely tied to the factory functions format. I'm not sure I can picture the alternative you're suggesting though, maybe worth an offline chat?